File: src/values/null.rs
use crate::bytecodes::ApicaTypeBytecode;
use crate::values::_type::ValueType;
use crate::values::bool::ValueBool;
use crate::values::char::ValueChar;
use crate::values::error::ValueError;
use crate::values::f32::ValueF32;
use crate::values::f64::ValueF64;
use crate::values::i16::ValueI16;
use crate::values::i32::ValueI32;
use crate::values::i64::ValueI64;
use crate::values::i8::ValueI8;
use crate::values::string::ValueString;
use crate::values::u16::ValueU16;
use crate::values::u32::ValueU32;
use crate::values::u64::ValueU64;
use crate::values::u8::ValueU8;
use crate::values::value::Value;

pub struct ValueNull {

}

impl ValueNull {
    pub fn init() -> ValueNull {
        ValueNull{}
    }
    
    pub fn show(&self, end: char) {
        print!("null{end}");
    }
    
    pub fn is_null(&self) -> bool {
        true
    }
    
    pub fn get_type_representation(&self) -> &str {
        "null"
    }
    
    pub fn not(&self) -> Value {
        Value::Bool(ValueBool::init_with(true))
    }

    pub fn convert(&'_ self, _: ApicaTypeBytecode) -> Option<Value> {
        None // null is AUTOMATICALLY converted
    }

    pub fn auto_convert(&self, to: ApicaTypeBytecode) -> Value {
        match to {
            ApicaTypeBytecode::Any => Value::Null(ValueNull::init()),
            ApicaTypeBytecode::Null => Value::Null(ValueNull::init()),
            ApicaTypeBytecode::I8 => Value::I8(ValueI8::init_empty()),
            ApicaTypeBytecode::I16 => Value::I16(ValueI16::init_empty()),
            ApicaTypeBytecode::I32 => Value::I32(ValueI32::init_empty()),
            ApicaTypeBytecode::I64 => Value::I64(ValueI64::init_empty()),
            ApicaTypeBytecode::U8 => Value::U8(ValueU8::init_empty()),
            ApicaTypeBytecode::U16 => Value::U16(ValueU16::init_empty()),
            ApicaTypeBytecode::U32 => Value::U32(ValueU32::init_empty()),
            ApicaTypeBytecode::U64 => Value::U64(ValueU64::init_empty()),
            ApicaTypeBytecode::F32 => Value::F32(ValueF32::init_empty()),
            ApicaTypeBytecode::F64 => Value::F64(ValueF64::init_empty()),
            ApicaTypeBytecode::Bool => Value::Bool(ValueBool::init_empty()),
            ApicaTypeBytecode::Char => Value::Char(ValueChar::init_empty()),
            ApicaTypeBytecode::String => Value::String(ValueString::init_empty()),
            ApicaTypeBytecode::Error => Value::Error(ValueError::init_empty()),
            ApicaTypeBytecode::Type => Value::Type(ValueType::init_empty()),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::bytecodes::ApicaTypeBytecode;
    use crate::values::null::ValueNull;
    use crate::values::value::Value;

    #[test]
    fn test_base() {
        let null = Value::Null(ValueNull::init());

        assert_eq!(true, null.is_null());
        assert_eq!(ApicaTypeBytecode::Null, null.get_kind());
        assert_eq!("null", null.get_type_representation());
    }

    #[test]
    fn test_auto_convert() {
        let null = Value::Null(ValueNull::init());

        let to_null = null.auto_convert(ApicaTypeBytecode::Null);
        assert_eq!(true, to_null.is_some());
        let to_null_unwrapped = to_null.unwrap();
        assert_eq!(ApicaTypeBytecode::Null, to_null_unwrapped.get_kind());
        assert_eq!(true, to_null_unwrapped.is_null());

        let to_i8 = null.auto_convert(ApicaTypeBytecode::I8);
        assert_eq!(true, to_i8.is_some());
        let to_i8_unwrapped = to_i8.unwrap();
        assert_eq!(ApicaTypeBytecode::I8, to_i8_unwrapped.get_kind());
        assert_eq!(true, to_i8_unwrapped.is_null());
    }
}

File: src/values/null.cpp
#include "values/bool.hpp"
#include "values/i8.hpp"
#include "values/i16.hpp"
#include "values/i32.hpp"
#include "values/i64.hpp"
#include "values/u8.hpp"
#include "values/u16.hpp"
#include "values/u32.hpp"
#include "values/u64.hpp"
#include "values/f32.hpp"
#include "values/f64.hpp"
#include "values/char.hpp"
#include "values/string.hpp"
#include "values/type.hpp"
#include "values/null.hpp"
#include "values/error.hpp"
#include "elements.hpp"

using namespace common::values;

ValueNull::ValueNull() {

}

void ValueNull::show(char end) const {
    std::cout << "null<>" << end;
}

bool ValueNull::isNull() const {
    return true;
}

std::string ValueNull::getTypeRepr() const {
    return "null";
}

common::bytecodes::ApicaTypeBytecode ValueNull::getKind() const {
    return common::bytecodes::ApicaTypeBytecode::Null;
}

std::optional<Value*> ValueNull::add(const Value *) const {
    return std::nullopt;
}

std::optional<Value*> ValueNull::increment() {
    return std::nullopt;
}

std::optional<Value*> ValueNull::subtract(const Value *) const {
    return std::nullopt;
}

std::optional<Value*> ValueNull::decrement() {
    return std::nullopt;
}

std::optional<Value*> ValueNull::unaryNot() const {
    return new ValueBool(true);
}

std::optional<Value*> ValueNull::convert(common::bytecodes::ApicaTypeBytecode) const {
    return std::nullopt; // null is AUTOMATICALLY converted
}

std::optional<Value*> ValueNull::autoConvert(common::bytecodes::ApicaTypeBytecode to) const {
    switch (to) {
        case common::bytecodes::ApicaTypeBytecode::Any:
        case common::bytecodes::ApicaTypeBytecode::Null:
            return new ValueNull();

        case common::bytecodes::ApicaTypeBytecode::I8:
            return new ValueI8();
        
        case common::bytecodes::ApicaTypeBytecode::I16:
            return new ValueI16();
        
        case common::bytecodes::ApicaTypeBytecode::I32:
            return new ValueI32();
        
        case common::bytecodes::ApicaTypeBytecode::I64:
            return new ValueI64();

        case common::bytecodes::ApicaTypeBytecode::U8:
            return new ValueU8();
        
        case common::bytecodes::ApicaTypeBytecode::U16:
            return new ValueU16();
        
        case common::bytecodes::ApicaTypeBytecode::U32:
            return new ValueU32();
        
        case common::bytecodes::ApicaTypeBytecode::U64:
            return new ValueU64();

        case common::bytecodes::ApicaTypeBytecode::F32:
            return new ValueF32();

        case common::bytecodes::ApicaTypeBytecode::F64:
            return new ValueF64();

        case common::bytecodes::ApicaTypeBytecode::Bool:
            return new ValueBool();

        case common::bytecodes::ApicaTypeBytecode::Char:
            return new ValueChar();

        case common::bytecodes::ApicaTypeBytecode::String:
            return new ValueString();

        case common::bytecodes::ApicaTypeBytecode::Error:
            return new ValueError();

        case common::bytecodes::ApicaTypeBytecode::Type:
            return new ValueType(common::bytecodes::ApicaTypeBytecode::Null);
        
        default: return std::nullopt;
    }
}